packages/agent/src/oidc-v2.ts

```diff
@@ -0,0 +1,689 @@
+ *
+ * @param codeVerifier `code_verifier` value generated e.g. from {@link generateRandomCodeVerifier}.
+ *
+ * @see {@link https://datatracker.ietf.org/doc/html/rfc7636#section-4.2 | RFC 7636, Client Creates the Code Challenge}
+ */
+async function deriveCodeChallenge(codeVerifier: Uint8Array) {
+  const codeChallengeu8a = await Sha256.digest({ data: codeVerifier });
+  const codeChallengeb64url =
+    Convert.uint8Array(codeChallengeu8a).toBase64Url();
+
+  return { codeChallengeu8a, codeChallengeb64url };
+}
+
+// TODO: when implementing pure OIDC split up the Web5 and OIDC params
+async function createAuthRequest(
+  options: RequireOnly<
+    HybridAuthRequest,
+    | 'code_challenge'
+    | 'code_challenge_method'
+    | 'client_id'
+    | 'scope'
+    | 'redirect_uri'
+    | 'permission_requests'
+  >
+) {
+  // Generate a random state value to associate the authorization request with the response.
+  const { randomStateu8a, ramdomStateb64url } = generateRandomState();
+
+  // Generate a random nonce value to associate the ID Token with the authorization request.
+  const nonce = await deriveNonceFromInput(randomStateu8a);
+
+  const requestObject: HybridAuthRequest = {
+    ...options,
+    nonce,
+    response_type : 'id_token',
+    response_mode : 'direct_post',
+    state         : ramdomStateb64url,
+  };
+
+  return requestObject;
+}
+
+async function encryptAuthRequest({
+  jwt,
+  codeChallenge,
+  nonce,
+}: {
+  jwt: string;
+  codeChallenge: Uint8Array;
+  nonce: Uint8Array;
+}) {
+  const protectedHeader = {
+    alg : 'dir',
+    cty : 'JWT',
+    enc : 'XC20P',
+    typ : 'JWT',
+  };
+  const additionalData = Convert.object(protectedHeader).toUint8Array();
+  const jwtu8a = Convert.string(jwt).toUint8Array();
+  const chacha = xchacha20poly1305(codeChallenge, nonce, additionalData);
+  const ciphertextAndTag = chacha.encrypt(jwtu8a);
+
+  /** The cipher output concatenates the encrypted data and tag
+   * so we need to extract the values for use in the JWE. */
+  const ciphertext = ciphertextAndTag.subarray(0, -16);
+  const authenticationTag = ciphertextAndTag.subarray(-16);
+
+  const compactJwe = [
+    Convert.object(protectedHeader).toBase64Url(),
+    '', // Empty string since there is no wrapped key.
+    Convert.uint8Array(nonce).toBase64Url(),
+    Convert.uint8Array(ciphertext).toBase64Url(),
+    Convert.uint8Array(authenticationTag).toBase64Url(),
+  ].join('.');
+
+  return compactJwe;
+}
+
+// todo: split up oidc and connect specific stuff in the future
+async function createResponseObject(
+  options: RequireOnly<
+    HybridAuthResponse,
+    'iss' | 'sub' | 'aud' | 'delegation_grants' | 'privateKeyJwks'
+  >
+) {
+  const currentTimeInSeconds = Math.floor(Date.now() / 1000);
+
+  const responseObject: HybridAuthResponse = {
+    ...options,
+    iat : currentTimeInSeconds,
+    exp : currentTimeInSeconds + 600, // Expires in 10 minutes.
+  };
+
+  return responseObject;
+}
+
+async function signJwt({
```

Comment:
Sanity: can we expose/reuse existing lib in web5 packages? e.g. `Jwt.sign()` in `credentials` package?

Reply:
Point taken I can look into replacing if it's a 1:1 replacement but it's worth pointing out that everything here works for Daniel's demo right now as is

Reply:
Sounds good, please create an issue and reference it in a TODO comment here so we don't loose track.

Reply:
issue made